model/flow/sealing_segment.go

```diff
@@ -108,49 +251,39 @@ func (builder *SealingSegmentBuilder) isValidHeight(block *Block) bool {
 	return block.Header.Height == builder.highest().Header.Height+1
 }
 
-// hasValidSeal returns true if highest block in the segment contains a seal for the lowest block
+// hasValidSeal returns true if the latest seal as of highest is for lowest.
 func (builder *SealingSegmentBuilder) hasValidSeal() bool {
 	lowestID := builder.lowest().ID()
-	highest := builder.highest()
-
-	// due to the fact that lowest is not always sealed
-	// by highest, if highest does not have any seals check
-	// that a valid ancestor does.
-	if len(highest.Payload.Seals) == 0 {
-		for i := len(builder.blocks) - 2; i >= 0; i-- {
-			// get first block that contains any seal
-			block := builder.blocks[i]
-			if len(block.Payload.Seals) == 0 {
-				continue
-			}
 
-			// check if block seals lowest
-			for _, seal := range block.Payload.Seals {
-				if seal.BlockID == lowestID {
-					return true
-				}
-			}
+	// due to the fact that lowest is not always sealed by highest,
+	// if highest does not have any seals check that a valid ancestor does.
+	for i := len(builder.blocks) - 1; i >= 0; i-- {
 
-			return false
+		// get first block that contains any seal
+		block := builder.blocks[i]
+		if len(block.Payload.Seals) == 0 {
+			continue
 		}
-	}
 
-	// highest contains seals check if seal for lowest exists
-	for _, seal := range highest.Payload.Seals {
-		if seal.BlockID == lowestID {
-			return true
+		// check if block seals lowest
+		for _, seal := range block.Payload.Seals {
+			if seal.BlockID == lowestID {
+				return true
+			}
 		}
```

Comment:
⚠️  Here we are assuming a seal that seals the lowest block is the last seal. It might not always be the case. For example:

The following is not a valid SealingSegment, though the last block contains a seal to the first block, but that seal (Seal_A) is not the last seal. The last seal (Seal_B) actually seals `B`.

`A <- B <- C <- D(Result_A, Result_B) <- E (Seal_A, Seal_B)`.

So we need to first find the last seal from the payload. Usually, the last seal is the last one in the seals payload. So should we _just_ check the last one: `block.Payload.Seals[len(block.Payload.Seals) - 1]` ? This would assume there is such ordering rule and has been checked by the protocol when adding this block.

Reply:
Good catch. Since all the blocks are already added to the segment at this point, I think we can just use `LatestSeals` to check the latest seal of the highest block here.

Reply:
https://github.com/onflow/flow-go/pull/1755/commits/0b4997831571d628bc033fa947618c353f667447